pkg/controller/direct/alloydb/instance_controller.go

```diff
@@ -0,0 +1,391 @@
+	} else {
+		req := &alloydbpb.CreateInstanceRequest{
+			Parent:     a.id.Parent().String(),
+			InstanceId: a.id.ID(),
+			Instance:   resource,
+		}
+		op, err := a.gcpClient.CreateInstance(ctx, req)
+		if err != nil {
+			log.V(2).Info("error creating instance", "name", a.id, "error", err)
+			return fmt.Errorf("creating instance %s: %w", a.id, err)
+		}
+		created, err = op.Wait(ctx)
+		if err != nil {
+			log.V(2).Info("error waiting instance creation", "name", a.id, "error", err)
+			return fmt.Errorf("instance %s waiting creation: %w", a.id, err)
+		}
+		log.V(2).Info("successfully created instance", "name", a.id)
+	}
+
+	status := AlloyDBInstanceStatus_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = a.id.AsExternalRef()
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+// Update updates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *instanceAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating instance", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	parsedActual := AlloyDBInstanceSpec_FromProto(mapCtx, a.actual)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	updatePaths, err := compareInstance(ctx, parsedActual, &a.desired.Spec)
+	if err != nil {
+		return err
+	}
+	desiredLabels := a.desired.GetObjectMeta().GetLabels()
+	desiredLabels["managed-by-cnrm"] = "true"
+	if !reflect.DeepEqual(a.actual.GetLabels(), desiredLabels) {
+		log.V(2).Info("'metadata.labels' field is updated (-old +new)", cmp.Diff(a.actual.GetLabels(), desiredLabels))
+		updatePaths = append(updatePaths, "labels")
+	}
+
+	if len(updatePaths) == 0 {
+		log.V(2).Info("no field needs update", "name", a.id)
+		return nil
+	}
+	updateMask := &fieldmaskpb.FieldMask{
+		Paths: updatePaths,
+	}
+	desiredPb := AlloyDBInstanceSpec_ToProto(mapCtx, &a.desired.DeepCopy().Spec)
+	desiredPb.Labels = desiredLabels
+	desiredPb.Name = a.id.String()
+	req := &alloydbpb.UpdateInstanceRequest{
+		UpdateMask: updateMask,
+		Instance:   desiredPb,
+	}
+	op, err := a.gcpClient.UpdateInstance(ctx, req)
+	if err != nil {
+		log.V(2).Info("error updating instance", "name", a.id, "error", err)
+		return fmt.Errorf("updating instance %s: %w", a.id, err)
+	}
+	updated, err := op.Wait(ctx)
+	if err != nil {
+		log.V(2).Info("error waiting instance update", "name", a.id, "error", err)
+		return fmt.Errorf("instance %s waiting update: %w", a.id, err)
+	}
+	log.V(2).Info("successfully updated instance", "name", a.id)
+
+	status := AlloyDBInstanceStatus_FromProto(mapCtx, updated)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	return updateOp.UpdateStatus(ctx, status, nil)
+}
+
+func compareInstance(ctx context.Context, actual, desired *krm.AlloyDBInstanceSpec) (updatePaths []string, err error) {
```

Comment:
Yes, two problems:
1. `CompareProtoMessage` compares all the proto fields, but AlloyDBInstance only supports a subset of the proto fields supported by the API now.
2. `CompareProtoMessage` doesn't understand the default values determined by the API code logic. e.g. not-yet supported field, `spec.observabilityConfig.trackWaitEvents`, has a default value `true`: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/6fac4a53f89a60ac9792c9ee35e90c1a0d9e2876/config/samples/resources/alloydbinstance/primary-instance/alloydb_v1beta1_alloydbinstance.yaml#L43

Because of the problems, `CompareProtoMessage` always return an error no matter if there is any KRM object update or not.

Reply:
Discussed offline. CompareProtoMessage compares the parsed desired pb with the actual pb, and the actual pb is the result of Find() (API GET). actual pb contains all the service-defaulted values of those unsupported fields and will result in a diff.

Because there are actually quite some unsupported fields, I think it'll be easier for us to compare the supported fields in Update() explicitly, and switch to CompareProtoMessage once we support most/all pb fields.

Reply:
Can you list which fields that have GCP defaulted values are not supported?   It means we may need to deal with them specially in other places